### ISSUE TITLE
Generics -> Type error not reported

### ISSUE THREAD

**Issue by fis-cz** [microsoft/TypeScript#33679]
Hi, 

I am not sure if this is a bug or just a TypeScript limitation (the same works in C# well). I was trying to make simple pipe framework while I wanted to ensure only correct data type filters can be added to the pipe based on output of the previous data type filter. I wanted to ensure this using generics during the transpile time, but unfortunately I was not able to get it work. Please see example bellow, I am afraid my dev / eng skills are not deep enough to explain what I wanted to achieve and where I can see the problem:

**TypeScript Version:**  Nightly

**Search Terms:** TypeScript Generics (walked through most opened issues)

**Code**

```ts
// Pipe plugin with generic input and output types
class A<input, output> {
	
	// Only plugins matching the current plugin output should be possible to be piped
	public pipe<output2>(plugin: A<output, output2>): A<output, output2> {
		return plugin;
	}
	
}

// Plugin B -> input: string, output: number
class B extends A<string, number> {
}

// Plugin C -> input: number, output: string
class C extends A<number, string> {
}

// Plugin D -> input: Date, output: Date
class D extends A<Date, Date> {
}

// Instances
const b: B = new B();
const c: C = new C();
const d: D = new D();

// Piping tests


// Intellisense correctly says the C plugin is type of A<number, string>
const h = b.pipe(c);

// Error expected
// Intellisense correctly says the 'i' plugin returned is type of A<number, Date>
// but d is type of D<Date, Date> and compiler does not report an error
const i = b.pipe(d);

// Error expected
const j = c.pipe(d);

// Error expected
const k = d.pipe(b);
```

**Expected behavior:**
Compiler returns: Type 'A<Date, Date>' is not assignable to 'A<number, Date>' error on line 35

**Actual behavior:**
No error returned (although type of d is correctly reported by the intellisense)

**Playground Link:**
[Playground](https://www.typescriptlang.org/play/?ts=Nightly#code/PTAEAUEsAcFNWgGwK4HNIDtQHdIBcALUVWDWAJ0gGNRNpk9QBDDAE1AHsH7G8BPOAGcAUFURNBg0AEEAPHQYAaTtwYA+UAG9hASF0hQAeQyI+CFOgxSAtkzxUCmVKELwqycuVKMkaTCrweUEECLkR2ACN4aA5JSAjEeDwOUCiEGFhWXXoE6nS4WS5AhgAmNQAKX0sALhlC1TxlIp4ygEpauWalAJaNbR0dLzwPLCrMAG5dAF9dYRnhA3ALfwAhUABaDQU8WsE8SgxUJobajGRrKPJRcUlQNdgADzxSVik5PYOj0DOLij654QLMBLPxYADCGy2GB4p3Ol2OxR2wX2TmuEikEMezzYb1kP3hyM+-3mi2WWAAIpDaNCGLVyXZYAiYaB6c80bdKViXrjWYyWQziYCDABJKx4FhUWAiKgcMWpWprAC831g2Du5VakxlcqotQhyrIarBGq1sr2oFYdNABtVLJNQuBMCcLileBEDtAoueiEQkEEpH9oBlnlgVDwpmCTD4UlcoAhYywfpcAngHAAZnV8RRlB8nGpRGbGERlREAHTQDLlKiaj0AUU8HHIoEecDDmSBnow3t9-qsbkbXjDEcEUZjBHgAHJIBPzKDQEMRplaDGU5wM3Is+RlLz8wYIgwLcvk3A1yzZLztwLmGwgxxrBXEk3WBwpd8OIwvDFyIwWM2G1dtXNSBrVSctK1YGsO3rchG2bB5W2eLJAMYAArECqDAuByggyYoP-OCEPbZDQAAaxA1hMNgcoIk1IA)

**Related Issues:**
No similar issues found

**Comment by AnyhowStep** [comment]
A simplified example of your problem,
```ts
interface Foo<T> {}

declare const fooNum : Foo<number>;
//Both are structurally the same, assignment allowed
const fooStr : Foo<string> = fooNum;

interface Bar<T> {
	t : T
}

declare const barNum : Bar<number>;
//Both are structurally different, assignment not allowed
const barStr : Bar<string> = barNum;
```
[Playground](https://www.typescriptlang.org/play/?ts=Nightly#code/JYOwLgpgTgZghgYwgAgGIHt0B4AqA+ZAbwF8AoUgEwgQBs4oUF0QBnMZGTAOQFcBbZAC40mLCH4AjaHgDcpAPTyAQujAALZPRRsoPBGB5Q4NGgE9k67XD4QANJpYtgAcxA3wmk+gDuECqSZWdk50AGUwKCERbB1QZwIAXg5ufjlSUEhYRBQlelwCQlIASHZhHFIySmo6BmRAtmQJel4BYVyoMUlpOUUVdU1anT0DIxNzCmAYGGgIcHs4Rxc3WfYQVU8aHz8A5gamqHDItrzYkHjkJP2WmSA)

**Comment by fis-cz** [comment]
Great! Thanks for explanation.
